src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+        this.pendingConnections.forEach((pending) => {
+            pending.reject(new ClientNotActiveError('Hazelcast client is shutting down'));
+        });
+
+        this.activeConnections.forEach((connection) => {
+           connection.close('Hazelcast client is shutting down', null);
+        });
+
+        this.removeAllListeners(CONNECTION_REMOVED_EVENT_NAME);
+        this.removeAllListeners(CONNECTION_ADDED_EVENT_NAME);
+        this.heartbeatManager.shutdown();
+    }
+
+    public getConnection(uuid: UUID): ClientConnection {
+        return this.activeConnections.get(uuid.toString());
+    }
+
+    public isInvocationAllowed(): Error {
+        const state = this.clientState;
+        if (state === ClientState.INITIALIZED_ON_CLUSTER && this.activeConnections.size > 0) {
+            return null;
+        }
+
+        let error: Error;
+        if (state === ClientState.INITIAL) {
+            if (this.asyncStart) {
+                error = new ClientOfflineError();
+            } else {
+                error = new IOError('No connection found to cluster since the client is starting.');
+            }
+        } else if (this.reconnectMode === ReconnectMode.ASYNC) {
+            error = new ClientOfflineError();
+        } else {
+            error = new IOError('No connection found to cluster.');
+        }
+        return error;
+    }
+
+    public getActiveConnections(): ClientConnection[] {
+        return Array.from(this.activeConnections.values());
+    }
+
+    public isAlive(): boolean {
+        return this.alive;
+    }
+
+    public getClientUuid(): UUID {
+        return this.clientUuid;
+    }
+
+    public getOrConnect(address: Address): Promise<ClientConnection> {
+        if (!this.client.getLifecycleService().isRunning()) {
+            return Promise.reject(new ClientNotActiveError('Client is not active.'));
+        }
+
+        const connection = this.getConnectionFromAddress(address);
+        if (connection) {
+            return Promise.resolve(connection);
+        }
+
+        const addressKey = address.toString();
+        const pendingConnection = this.pendingConnections.get(addressKey);
+        if (pendingConnection) {
+            return pendingConnection.promise;
+        }
+
+        const connectionResolver: Promise.Resolver<ClientConnection> = DeferredPromise<ClientConnection>();
+        this.pendingConnections.set(addressKey, connectionResolver);
+
+        const processResponseCallback = (msg: ClientMessage) => {
+            this.client.getInvocationService().processResponse(msg);
+        };
+
+        this.translate(address)
+            .then((translatedAddress) => {
+                if (translatedAddress == null) {
+                    connectionResolver.reject(new RangeError(`Address Translator could not translate address ${address}`));
+                    return;
+                }
+
+                this.triggerConnect(translatedAddress)
+                    .then((socket) => {
```

Comment:
nit: nested `then()` calls should be avoided, when possible, as this helps readability.

Reply:
There are variables that I wanted these closures to capture. So, that's way they are nested. I could define these variables in the main function body and use it. Should I do it ?

Reply:
I don't mind extracting a couple of variables from the chain. Here is how the code might look in this case.

Please note that I've also removed `finally` block in the snippet, but this change depends on the discussion in https://github.com/hazelcast/hazelcast-nodejs-client/pull/522#discussion_r439492747.

```js
        let translatedAddress: Address;
        let clientConnection: ClientConnection;
        this.translate(address)
            .then((translated) => {
                translatedAddress = translated;
                if (translatedAddress == null) {
                    throw new RangeError(`Address Translator could not translate address ${address}`);
                }
                return this.triggerConnect(translatedAddress);
            })
            .then((socket) => {
                clientConnection =
                    new ClientConnection(this.client, translatedAddress, socket, this.connectionIdCounter++);
                return this.initiateCommunication(socket)
            })
            .then(() => clientConnection.registerResponseCallback(processResponseCallback))
            .then(() => this.authenticateOnCluster(clientConnection))
            .then(() => connectionResolver.resolve(clientConnection))
            .catch((error) => connectionResolver.reject(error));
```

I find this code a bit easier to read than the original one, even despite of those "external" variables. But it's certainly a matter of taste, so it's up to you to make the decision.

Reply:
In fact, this looks much better. I replaced this part with the code you posted